File: 739-daily-temperatures/daily-temperatures.cpp
class Solution {
public:
    // temps , answer -> ans[i] number of days you have to wait after ith day to get warmer tmep; 
    // for each i loop over i+1-> end to find the warmer temp
    vector<int> dailyTemperatures(vector<int>& t) {
        stack<int> st;
        vector<int> vv;  
        for(int i = t.size()-1; i >=0; i --){
            while(!st.empty() && t[st.top()] <= t[i]) {
                st.pop();
            }

            if(st.empty()) {
                vv.push_back(0);
            }else{
                vv.push_back(st.top() - i);
            }
            st.push(i);
        }
        reverse(vv.begin(), vv.end());
        return vv; 
    }
};

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, t: List[int]) -> List[int]:
        st = deque([])

        # next greater element.
        arr = [0]*len(t) 

        for i in range(len(t)-1, -1, -1):
            if not st :
                arr[i] = 0
            elif st[-1][0] <= t[i]:
                print(st[-1][0], t[i])
                while st and st[-1][0] <= t[i]:
                    st.pop()
                if st:
                    arr[i] = st[-1][1] - i
                else :
                    arr[i] = 0
            else :
                arr[i] = st[-1][1] - i
        
            st.append([t[i], i])

        return arr  





        